File: examples/backend/aspnet-core/Services/UserService.cs
// ---------------------------------------------------------------------------
// UserService.cs
// Service that extracts user information from the ClaimsPrincipal.
//
// Provides a clean API for Razor Pages to access user details such as
// name, email, and roles without directly parsing claims.
// ---------------------------------------------------------------------------

using System.Security.Claims;

namespace IAMExample.Services;

/// <summary>
/// Provides helpers for extracting user information from the current
/// <see cref="ClaimsPrincipal"/>.
/// </summary>
public class UserService
{
    /// <summary>
    /// Gets the display name of the authenticated user.
    /// Falls back through preferred_username, name, and email claims.
    /// </summary>
    /// <param name="user">The claims principal.</param>
    /// <returns>The user's display name, or an empty string if unauthenticated.</returns>
    public string GetDisplayName(ClaimsPrincipal user)
    {
        return user.FindFirstValue("preferred_username")
            ?? user.FindFirstValue(ClaimTypes.Name)
            ?? user.FindFirstValue("name")
            ?? user.FindFirstValue(ClaimTypes.Email)
            ?? string.Empty;
    }

    /// <summary>
    /// Gets the email address of the authenticated user.
    /// </summary>
    /// <param name="user">The claims principal.</param>
    /// <returns>The email address, or null if not available.</returns>
    public string? GetEmail(ClaimsPrincipal user)
    {
        return user.FindFirstValue(ClaimTypes.Email)
            ?? user.FindFirstValue("email");
    }

    /// <summary>
    /// Gets the full name of the authenticated user.
    /// </summary>
    /// <param name="user">The claims principal.</param>
    /// <returns>The full name, or null if not available.</returns>
    public string? GetFullName(ClaimsPrincipal user)
    {
        return user.FindFirstValue("name")
            ?? user.FindFirstValue(ClaimTypes.Name);
    }

    /// <summary>
    /// Gets the list of realm roles assigned to the user.
    /// </summary>
    /// <param name="user">The claims principal.</param>
    /// <returns>A list of role names.</returns>
    public IReadOnlyList<string> GetRoles(ClaimsPrincipal user)
    {
        return user.FindAll(ClaimTypes.Role)
            .Select(c => c.Value)
            .Distinct()
            .OrderBy(r => r)
            .ToList();
    }

    /// <summary>
    /// Checks whether the user has the admin role.
    /// </summary>
    /// <param name="user">The claims principal.</param>
    /// <returns>True if the user has the admin role.</returns>
    public bool IsAdmin(ClaimsPrincipal user)
    {
        return user.HasClaim(ClaimTypes.Role, "admin");
    }

    /// <summary>
    /// Gets all claims as key-value pairs for display purposes.
    /// </summary>
    /// <param name="user">The claims principal.</param>
    /// <returns>A list of claim type/value pairs.</returns>
    public IReadOnlyList<(string Type, string Value)> GetAllClaims(ClaimsPrincipal user)
    {
        return user.Claims
            .Select(c => (Type: c.Type, Value: c.Value))
            .ToList();
    }
}

File: examples/backend/quarkus/src/main/java/com/ximplicity/iam/service/UserService.java
package com.ximplicity.iam.service;

import io.quarkus.oidc.UserInfo;
import io.quarkus.security.identity.SecurityIdentity;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import org.eclipse.microprofile.jwt.JsonWebToken;

import java.util.Collections;
import java.util.Map;
import java.util.Optional;
import java.util.Set;

/**
 * Service that extracts user information from the current
 * {@link SecurityIdentity} and OIDC token claims.
 *
 * <p>This centralises all user-data extraction so that resource classes
 * do not need to interact with the raw JWT or security identity
 * directly.</p>
 */
@ApplicationScoped
public class UserService {

    private final SecurityIdentity identity;
    private final JsonWebToken jwt;

    /**
     * Constructs a new {@code UserService}.
     *
     * @param identity the current security identity (injected per-request)
     * @param jwt      the parsed JSON Web Token (injected per-request)
     */
    @Inject
    public UserService(SecurityIdentity identity, JsonWebToken jwt) {
        this.identity = identity;
        this.jwt = jwt;
    }

    /**
     * Returns the preferred username, falling back to the JWT subject.
     *
     * @return the display name for the current user
     */
    public String getDisplayName() {
        String preferred = jwt.getClaim("preferred_username");
        if (preferred != null && !preferred.isBlank()) {
            return preferred;
        }
        String name = jwt.getClaim("name");
        if (name != null && !name.isBlank()) {
            return name;
        }
        return Optional.ofNullable(jwt.getSubject()).orElse("Unknown");
    }

    /**
     * Returns the user's full name from the token.
     *
     * @return the full name, or "Not provided" if absent
     */
    public String getFullName() {
        String name = jwt.getClaim("name");
        return (name != null && !name.isBlank()) ? name : "Not provided";
    }

    /**
     * Returns the user's email address from the token.
     *
     * @return the email, or "Not provided" if absent
     */
    public String getEmail() {
        String email = jwt.getClaim("email");
        return (email != null && !email.isBlank()) ? email : "Not provided";
    }

    /**
     * Returns the set of roles assigned to the current user.
     *
     * @return an unmodifiable set of role names
     */
    public Set<String> getRoles() {
        return Collections.unmodifiableSet(identity.getRoles());
    }

    /**
     * Checks whether the current user has a specific role.
     *
     * @param role the role name to check
     * @return {@code true} if the user holds the role
     */
    public boolean hasRole(String role) {
        return identity.hasRole(role);
    }

    /**
     * Returns the JWT subject (sub claim).
     *
     * @return the subject identifier
     */
    public String getSubject() {
        return Optional.ofNullable(jwt.getSubject()).orElse("N/A");
    }

    /**
     * Returns the JWT issuer (iss claim).
     *
     * @return the issuer URL
     */
    public String getIssuer() {
        return Optional.ofNullable(jwt.getIssuer()).orElse("N/A");
    }

    /**
     * Returns the token expiration time as an epoch second, or -1 if
     * unavailable.
     *
     * @return expiration epoch seconds
     */
    public long getTokenExpiration() {
        return jwt.getExpirationTime();
    }

    /**
     * Returns whether the current identity is authenticated (not anonymous).
     *
     * @return {@code true} if the user is logged in
     */
    public boolean isAuthenticated() {
        return !identity.isAnonymous();
    }

    /**
     * Returns the raw access token string for display/preview purposes.
     *
     * @return the raw token string, or an empty string
     */
    public String getRawToken() {
        return Optional.ofNullable(jwt.getRawToken()).orElse("");
    }
}
